services/src/main/java/org/keycloak/services/clienttype/impl/DefaultClientTypeProviderFactory.java

```diff
@@ -0,0 +1,96 @@
+/*
+ * Copyright 2021 Red Hat, Inc. and/or its affiliates
+ * and other contributors as indicated by the @author tags.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *
+ */
+
+package org.keycloak.services.clienttype.impl;
+
+import java.beans.BeanInfo;
+import java.beans.IntrospectionException;
+import java.beans.Introspector;
+import java.beans.PropertyDescriptor;
+import java.lang.reflect.Method;
+import java.util.Arrays;
+import java.util.Map;
+import java.util.Set;
+import java.util.function.Function;
+import java.util.stream.Collectors;
+
+import org.keycloak.Config;
+import org.keycloak.models.KeycloakSession;
+import org.keycloak.models.KeycloakSessionFactory;
+import org.keycloak.representations.idm.ClientRepresentation;
+import org.keycloak.client.clienttype.ClientTypeProvider;
+import org.keycloak.client.clienttype.ClientTypeProviderFactory;
+
+/**
+ * @author <a href="mailto:mposolda@redhat.com">Marek Posolda</a>
+ */
+public class DefaultClientTypeProviderFactory implements ClientTypeProviderFactory {
```

Comment:
Is it an option to implement `EnvironmentDependentFactory` and make this provider working just if `Client types` feature is enabled? See other providers relying on the "feature" being enabled.

Reply:
I think I implemented it correctly, but please feel free to let me know if there was more I was supposed to add. I also cleaned up some of the TODOs, but I just took your name off the doc ones and the ones that I want to get my team's input on.

Reply:
I think it will be good if you also implement the method `isSupported()`, which will make this provider supported just if "client types" feature is supported.

I think something like this:
```
@OVerride
public boolean isSupported(Config.Scope config) {
    return Profile.isFeatureEnabled(Profile.Feature.CLIENT_TYPES);
}
```

For the inspiration, see for example class `WebAuthnAuthenticatorFactory`, which does something like this for WebAuthn feature.

Reply:
Yeah, I ran into issues in my next MR with client types not working so its been added there, but I'll put it into this MR too. Thank you for taking another look!